ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/TransactionsReader.scala

```diff
@@ -366,35 +365,19 @@ private[appendonlydao] final class TransactionsReader(
       Telemetry.Transactions.createSpan(activeAt)(qualifiedNameOfCurrentFunc)
     logger.debug(s"getActiveContracts($activeAt, $filter, $verbose)")
 
-    val requestedRangeF: Future[EventsRange[(Offset, Long)]] = getAcsEventSeqIdRange(activeAt)
-
-    val query = (range: EventsRange[(Offset, Long)]) => { implicit connection: Connection =>
-      logger.debug(s"getActiveContracts query($range)")
-      queryNonPruned.executeSql(
-        getActiveContracts(
-          range,
-          filter,
-          pageSize,
-        )(connection),
-        activeAt,
-        pruned =>
-          s"Active contracts request after ${activeAt.toHexString} precedes pruned offset ${pruned.toHexString}",
+    Source
+      .futureSource(
+        getAcsEventSeqIdRange(activeAt)
+          .map(requestedRange => acsReader.acsStream(filter, requestedRange.endInclusive))
       )
-    }
-
-    val events: Source[EventsTable.Entry[Event], NotUsed] =
-      Source
-        .futureSource(requestedRangeF.map { requestedRange =>
-          streamEvents(
-            verbose,
-            dbMetrics.getActiveContracts,
-            query,
-            nextPageRange[Event](requestedRange.endInclusive),
-          )(requestedRange)
-        })
-        .mapMaterializedValue(_ => NotUsed)
-
-    groupContiguous(events)(by = _.transactionId)
+      .mapAsync(eventProcessingParallelism) { rawResult =>
+        Timed.future(
+          future = Future.traverse(rawResult)(
```

Comment:
The parallelism of the surrounding `mapAsync` is bounded by `eventProcessingParallelism`, but the parallelism of `Future.traverse` is (I assume) unbounded. Did you check whether the `mapAsync` has any measurable effect on performance?

You could also move the call to `EventsTable.Entry.toGetActiveContractsResponse` to the `.mapAsync` block, and use `.mapConcat(identity)` below, for _slightly_ less sequential processing.

Reply:
hm it seems to me the deserializeEntry is doing
* all deserialization
* all enriching/Engine execution
BEFORE the future is created, so these things not only not running in parallel, but not even in a future(see traverse code: it executes the future creating function sequentially), which practically eliminates any good effect of the mapAsync.
It looks like that only the package loading is running in a real future. So we seem to have unbounded parallelism for package loading here :thinking: 
If that analyses is true, we need to wrap the creation of the future into a future to at least have parallelism of 2 here.

Hm it is actually quite nasty because it does all this deserializing computation on a thread which should just move around stuff in akka-streams...I am wondering how malicious this is to stream performance....oh my, we might have found the throughput cap I was talking about before :metal:

Reply:
Good point, most of the work of `deserializeEntry` is executed before the first Future is created.